x/cronos/keeper/grpc_query.go

```diff
@@ -40,3 +42,68 @@ func (k Keeper) DenomByContract(goCtx context.Context, req *types.DenomByContrac
 		Denom: denom,
 	}, nil
 }
+
+// ReplayBlock replay the eth messages in the block to recover the results of false-failed txs.
+func (k Keeper) ReplayBlock(goCtx context.Context, req *types.ReplayBlockRequest) (*types.ReplayBlockResponse, error) {
+	rsps := make([]*evmtypes.MsgEthereumTxResponse, 0, len(req.Msgs))
+
+	// prepare the block context, the multistore version should be setup already in grpc query context.
+	ctx := sdk.UnwrapSDKContext(goCtx).
+		WithBlockHeight(req.BlockNumber).
+		WithBlockTime(req.BlockTime).
+		WithHeaderHash(common.Hex2Bytes(req.BlockHash))
+
+	// load parameters
+	params := k.evmKeeper.GetParams(ctx)
+	chainID := k.evmKeeper.ChainID()
+	// the chain_id is irrelevant here
+	ethCfg := params.ChainConfig.EthereumConfig(chainID)
+
+	blockHeight := big.NewInt(req.BlockNumber)
+	homestead := ethCfg.IsHomestead(blockHeight)
+	istanbul := ethCfg.IsIstanbul(blockHeight)
+	london := ethCfg.IsLondon(blockHeight)
+	evmDenom := params.EvmDenom
+
+	// we assume the message executions are successful, they are filtered in json-rpc api
+	for _, msg := range req.Msgs {
+		// deduct fee
+		txData, err := evmtypes.UnpackTxData(msg.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		// populate the `From` field
+		if _, err := msg.GetSender(chainID); err != nil {
+			return nil, err
+		}
+
+		if _, err := k.evmKeeper.DeductTxCostsFromUserBalance(
+			ctx,
+			*msg,
+			txData,
+			evmDenom,
+			homestead,
+			istanbul,
+			london,
+		); err != nil {
+			return nil, err
+		}
+
+		// increase nonce
+		acc := k.accountKeeper.GetAccount(ctx, msg.GetFrom())
```

Comment:
I think there should verify the acc not equal to nil

Reply:
done, added a `nil` check.